infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/KubernetesServerExposer.java

```diff
@@ -175,57 +172,89 @@ public void expose(Map<String, ? extends ServerConfig> servers) throws Infrastru
             internalServers.put(key, value);
+            servicePorts.put(value.getPort(), sp);
           } else {
             // Server is external. Check if it should be secure or not
             if (value.isSecure()) {
               secureServers.put(key, value);
+              securePorts.put(value.getPort(), sp);
             } else {
               externalServers.put(key, value);
+              servicePorts.put(value.getPort(), sp);
             }
           }
         });
 
-    Collection<ServicePort> servicePorts = exposePorts(servers.values());
-    Service service =
-        new ServerServiceBuilder()
-            .withName(generate(SERVER_PREFIX, SERVER_UNIQUE_PART_SIZE) + '-' + machineName)
-            .withMachineName(machineName)
-            .withSelectorEntry(CHE_ORIGINAL_NAME_LABEL, pod.getMetadata().getName())
-            .withPorts(new ArrayList<>(servicePorts))
-            .withServers(internalServers)
-            .build();
+    Map<String, ServerConfig> allNonSecureServers = new HashMap<>(internalServers);
+    allNonSecureServers.putAll(externalServers);
+    // only explicitly expose the service for the non-secure servers
+    if (!servicePorts.isEmpty()) {
+      Service service =
+          new ServerServiceBuilder()
+              .withName(generate(SERVER_PREFIX, SERVER_UNIQUE_PART_SIZE) + '-' + machineName)
+              .withMachineName(machineName)
+              .withSelectorEntry(CHE_ORIGINAL_NAME_LABEL, pod.getMetadata().getName())
+              .withPorts(new ArrayList<>(servicePorts.values()))
+              .withServers(allNonSecureServers)
+              .build();
 
-    String serviceName = service.getMetadata().getName();
-    k8sEnv.getServices().put(serviceName, service);
+      String serviceName = service.getMetadata().getName();
+      k8sEnv.getServices().put(serviceName, service);
 
-    for (ServicePort servicePort : servicePorts) {
-      // expose service port related external servers if exist
-      Map<String, ServerConfig> matchedExternalServers = match(externalServers, servicePort);
-      if (!matchedExternalServers.isEmpty()) {
-        onEachExposableServerSet(
-            matchedExternalServers,
-            (serverId, srvrs) -> {
-              externalServerExposer.expose(
-                  k8sEnv, machineName, serviceName, serverId, servicePort, srvrs);
-            });
+      for (ServicePort servicePort : servicePorts.values()) {
+        // expose service port related external servers if exist
+        Map<String, ServerConfig> matchedExternalServers = match(externalServers, servicePort);
+        if (!matchedExternalServers.isEmpty()) {
+          onEachExposableServerSet(
+              matchedExternalServers,
+              (serverId, srvrs) -> {
+                externalServerExposer.expose(
+                    k8sEnv, machineName, serviceName, serverId, servicePort, srvrs);
+              });
+        }
+      }
+    }
+
+    if (!secureServers.isEmpty()) {
```

Comment:
can we extract this block to separate method? It's very hard to follow such long method doing so many things